File: src/TemporalOps.cpp
#include "../include/TemporalOps.h"
#include <algorithm>

using namespace cv;

std::vector<Mat> TemporalOps::reverse(const std::vector<Mat>& frames) {
    std::vector<Mat> result = frames;
    std::reverse(result.begin(), result.end());
    return result;
}

std::vector<Mat> TemporalOps::speedUp(const std::vector<Mat>& frames, double factor) {
    std::vector<Mat> result;
    int step = (int)factor;
    if (step < 1) step = 1;
    for (size_t i = 0; i < frames.size(); i += step) result.push_back(frames[i].clone());
    return result;
}

std::vector<Mat> TemporalOps::freeze(const std::vector<Mat>& frames, int frameIdx, int duration) {
    if (frames.empty() || frameIdx >= (int)frames.size()) return frames;
    std::vector<Mat> result;
    for (int i = 0; i < frameIdx; i++) result.push_back(frames[i].clone());
    Mat freezeFrame = frames[frameIdx].clone();
    for (int i = 0; i < duration; i++) result.push_back(freezeFrame.clone());
    for (size_t i = frameIdx; i < frames.size(); i++) result.push_back(frames[i].clone());
    return result;
}

std::vector<Mat> TemporalOps::loop(const std::vector<Mat>& frames, int count) {
    std::vector<Mat> result;
    for (int c = 0; c < count; c++) for (const auto& f : frames) result.push_back(f.clone());
    return result;
}

std::vector<Mat> TemporalOps::duplicate(const std::vector<Mat>& frames, int factor) {
    std::vector<Mat> result;
    for (const auto& f : frames) for (int i = 0; i < factor; ++i) result.push_back(f.clone());
    return result;
}

File: src/videts/temporal_ops.py
class TemporalOps:
    @staticmethod
    def reverse(frames):
        return frames[::-1]
    
    @staticmethod
    def speed_up(frames, factor):
        step = int(factor)
        return frames[::step]
    
    @staticmethod
    def slow_down(frames):
        return frames
    
    @staticmethod
    def freeze(frames, frame_idx, duration):
        if not frames or frame_idx >= len(frames):
            return frames
        freeze_frame = frames[frame_idx]
        result = frames[:frame_idx]
        for _ in range(duration):
            result.append(freeze_frame.copy())
        result.extend(frames[frame_idx:])
        return result
    
    @staticmethod
    def loop(frames, count):
        return frames * count
    
    @staticmethod
    def duplicate(frames, factor):
        result = []
        for f in frames:
            for _ in range(factor):
                result.append(f.copy())
        return result
